### ISSUE TITLE
APK tool error

### ISSUE THREAD

**Issue by theamanrawat** [iBotPeaches/Apktool#2259]
### Information
1. **Apktool Version (`apktool -version`)** - 2.4.1
2. **Operating System (Mac, Linux, Windows)** - Window 10
3. **APK From? (Playstore, ROM, Other)** -Playstore

### Stacktrace/Logcat

```
I: Using Apktool 2.4.1 on BOB_mPassbook_v3.0.7.apk
I: Loading resource table...
I: Decoding AndroidManifest.xml with resources...
I: Loading resource table from file: C:\Users\rawat\AppData\Local\apktool\framework\1.apk
I: Regular manifest package...
I: Decoding file-resources...
W: Cant find 9patch chunk in file: "drawable-hdpi-v4/common_google_signin_btn_icon_light_normal_background.9.png". Renaming it to *.png.
W: Cant find 9patch chunk in file: "drawable-xhdpi-v4/common_google_signin_btn_text_light_normal_background.9.png". Renaming it to *.png.
W: Cant find 9patch chunk in file: "drawable-xxhdpi-v4/common_google_signin_btn_text_light_normal_background.9.png". Renaming it to *.png.
W: Cant find 9patch chunk in file: "drawable-xxhdpi-v4/drawer_shadow.9.png". Renaming it to *.png.
W: Cant find 9patch chunk in file: "drawable-hdpi-v4/common_google_signin_btn_text_dark_normal_background.9.png". Renaming it to *.png.
W: Cant find 9patch chunk in file: "drawable-xhdpi-v4/drawer_shadow.9.png". Renaming it to *.png.
W: Cant find 9patch chunk in file: "drawable-mdpi-v4/common_google_signin_btn_icon_dark_normal_background.9.png". Renaming it to *.png.
W: Cant find 9patch chunk in file: "drawable-mdpi-v4/common_google_signin_btn_text_dark_normal_background.9.png". Renaming it to *.png.
W: Cant find 9patch chunk in file: "drawable-xhdpi-v4/common_google_signin_btn_text_dark_normal_background.9.png". Renaming it to *.png.
W: Cant find 9patch chunk in file: "drawable-hdpi-v4/sigup_bg.9.png". Renaming it to *.png.
W: Cant find 9patch chunk in file: "drawable-mdpi-v4/common_google_signin_btn_icon_light_normal_background.9.png". Renaming it to *.png.
W: Cant find 9patch chunk in file: "drawable-xxhdpi-v4/common_google_signin_btn_icon_dark_normal_background.9.png". Renaming it to *.png.
W: Cant find 9patch chunk in file: "drawable-mdpi-v4/drawer_shadow.9.png". Renaming it to *.png.
W: Cant find 9patch chunk in file: "drawable-mdpi-v4/common_google_signin_btn_text_light_normal_background.9.png". Renaming it to *.png.
W: Cant find 9patch chunk in file: "drawable-xhdpi-v4/common_google_signin_btn_icon_dark_normal_background.9.png". Renaming it to *.png.
W: Cant find 9patch chunk in file: "drawable-hdpi-v4/common_google_signin_btn_icon_dark_normal_background.9.png". Renaming it to *.png.
W: Cant find 9patch chunk in file: "drawable-xhdpi-v4/common_google_signin_btn_icon_light_normal_background.9.png". Renaming it to *.png.
W: Cant find 9patch chunk in file: "drawable-xxhdpi-v4/common_google_signin_btn_icon_light_normal_background.9.png". Renaming it to *.png.
W: Cant find 9patch chunk in file: "drawable-hdpi-v4/drawer_shadow.9.png". Renaming it to *.png.
W: Cant find 9patch chunk in file: "drawable-hdpi-v4/common_google_signin_btn_text_light_normal_background.9.png". Renaming it to *.png.
W: Cant find 9patch chunk in file: "drawable-xxhdpi-v4/common_google_signin_btn_text_dark_normal_background.9.png". Renaming it to *.png.
I: Decoding values */* XMLs...
I: Baksmaling classes.dex...
I: Copying assets and libs...
I: Copying unknown files...
I: Copying original files...
```

### Steps to Reproduce
1. `apktool d app.apk`
1. `apktool b app.apk`

### Frameworks
If this APK is from an OEM ROM (Samsung, HTC, LG). Please attach framework files
(`.apks` that live in `/system/framework` or `/system/priv-app`)

### APK
If this APK can be freely shared, please upload/attach a link to it.

### Questions to ask before submission
1. Have you tried `apktool d`, `apktool b` without changing anything?
>Yes
3. Are you using the latest apktool version?
>Yes

**Comment by theamanrawat** [comment]
apktool is not able to rebuild most of the apps I have tried with many apps but it is giving different kind of errors every times.

**Comment by JimDunn1234** [comment]
Agreed, apktool has major issues with png files now or -v4 paths in res folder. Its always the -v4 png's I have noticed.

**Comment by iBotPeaches** [comment]
Do you have an example apk? Our test suite is pretty powerful with 9patch images, so confused how this may have regressed. 

@JimDunn1234 - You seem to have some serious negativity around this project. Its open source so you can contribute :)

**Comment by theamanrawat** [comment]
I tested Instagram app

**Comment by Ticklefish** [comment]
If you can share the APK you're having trouble with, other people can take a look at it and hopefully give you an answer.
To me, this just looks the 9patch images are malformed. Which does happen occasionally. If that's the case, it doesn't mean the APK won't work. The worst that will happen is some of the images might look a bit strange.

**Comment by aktivx** [comment]
same error apktool 2.4.1 with sky_go_pr17-5-0-1-110
tried 2.3.4 also but no luck

you can get the apk from here
https://www.apkmirror.com/apk/sky-italia-s-r-l/sky-go-it/sky-go-it-pr17-5-0-1-1100-release/

**Comment by Ticklefish** [comment]
Thanks for sharing the APK. I downloaded that particular version and did some experimenting with it, after installing a framework-res.apk from an AOSP build of Android 10 (sigh...I really do miss Google giving names to these - 'Android 10' just sounds so dull..).

This is what I get when I decompile:

> (java -Xmx1024m -jar user_files\APKTOOL\apktool_2.4.1.jar d --force-manifest     --only-main-classes  -f -p _WorkArea1\Frameworks _WorkArea1\_in\it.sky.anywhere.apk -o _WorkArea1\_working\it.sky.anywhere.apk) 
> 
> I: Using Apktool 2.4.1 on it.sky.anywhere.apk
> I: Loading resource table...
> I: Decoding AndroidManifest.xml with resources...
> I: Loading resource table from file: _WorkArea1\Frameworks\1.apk
> I: Regular manifest package...
> I: Decoding file-resources...
> W: Cant find 9patch chunk in file: "drawable-mdpi-v4/grid_item_border.9.png". Renaming it to *.png.
> W: Cant find 9patch chunk in file: "drawable-xhdpi-v4/grid_item_border.9.png". Renaming it to *.png.
> W: Cant find 9patch chunk in file: "drawable-hdpi-v4/grid_item_border.9.png". Renaming it to *.png.
> W: Cant find 9patch chunk in file: "drawable-xxhdpi-v4/grid_item_border.9.png". Renaming it to *.png.
> I: Decoding values */* XMLs...
> I: Baksmaling classes.dex...
> I: Baksmaling classes2.dex...
> I: Copying assets and libs...
> I: Copying unknown files...
> I: Copying original files...
> I: Copying META-INF/services directory

Four 9patch ("***.9.PNG**") files are missing the actual 9patch information so apktool is turning them into ordinary PNG files. This isn't that unusual, to be honest. I've seen this a lot over the years and it's normally nothing to worry about. Since those files didn't have 9patch information in them you won't see any difference visually between them and the PNG versions when the app is running. There might be some difference in file sizes possibly but nothing significant. This can be ignored.

So I tried recompiling. First with aapt2 enabled:

> (java -jar user_files\APKTOOL\apktool_2.4.1.jar b --copy-original  --force-all --use-aapt2 -p _WorkArea1\Frameworks _WorkArea1\_working\it.sky.anywhere.apk -o _WorkArea1\_out\it.sky.anywhere.apk) 
> 
> -c/--copy-original has been deprecated. Removal planned for v2.5.0 (#2129)
> I: Using Apktool 2.4.1
> I: Smaling smali folder into classes.dex...
> I: Smaling smali_classes2 folder into classes2.dex...
> I: Building resources...
> W: C:\TickleMyAndroid\_WorkArea1\_working\it.sky.anywhere.apk\res\values\public.xml:1169: error: resource 'drawable/$avd_hide_password__0' has invalid entry name '$avd_hide_password__0'. Invalid character '$avd_hide_password__0'.
> W: C:\TickleMyAndroid\_WorkArea1\_working\it.sky.anywhere.apk\res\values\public.xml:1170: error: resource 'drawable/$avd_hide_password__1' has invalid entry name '$avd_hide_password__1'. Invalid character '$avd_hide_password__1'.
> W: C:\TickleMyAndroid\_WorkArea1\_working\it.sky.anywhere.apk\res\values\public.xml:1171: error: resource 'drawable/$avd_hide_password__2' has invalid entry name '$avd_hide_password__2'. Invalid character '$avd_hide_password__2'.
> W: C:\TickleMyAndroid\_WorkArea1\_working\it.sky.anywhere.apk\res\values\public.xml:1172: error: resource 'drawable/$avd_show_password__0' has invalid entry name '$avd_show_password__0'. Invalid character '$avd_show_password__0'.
> W: C:\TickleMyAndroid\_WorkArea1\_working\it.sky.anywhere.apk\res\values\public.xml:1173: error: resource 'drawable/$avd_show_password__1' has invalid entry name '$avd_show_password__1'. Invalid character '$avd_show_password__1'.
> W: C:\TickleMyAndroid\_WorkArea1\_working\it.sky.anywhere.apk\res\values\public.xml:1174: error: resource 'drawable/$avd_show_password__2' has invalid entry name '$avd_show_password__2'. Invalid character '$avd_show_password__2'.
> I: Copying libs... (/lib)
> I: Copying libs... (/kotlin)
> I: Copying libs... (/META-INF/services)
> I: Copy original files...
> I: Copy AndroidManifest.xml...
> I: Copy META-INF...
> I: Building apk file...
> I: Copying unknown files/dir...
> I: Built apk...

This error about the invalid character is interesting. The '**$**' is causing this. Aapt2 doesn't like it. You can get around this by renaming every instance of those files to something else (call them "**dollarsign_avd..**" for example) which would work. Or you could just use aapt1 instead:

> (java -jar user_files\APKTOOL\apktool_2.4.1.jar b --copy-original  --force-all  -p _WorkArea1\Frameworks _WorkArea1\_working\it.sky.anywhere.apk -o _WorkArea1\_out\it.sky.anywhere.apk) 
> 
> -c/--copy-original has been deprecated. Removal planned for v2.5.0 (#2129)
> I: Using Apktool 2.4.1
> I: Smaling smali folder into classes.dex...
> I: Smaling smali_classes2 folder into classes2.dex...
> I: Building resources...
> I: Copying libs... (/lib)
> I: Copying libs... (/kotlin)
> I: Copying libs... (/META-INF/services)
> I: Copy original files...
> I: Copy AndroidManifest.xml...
> I: Copy META-INF...
> I: Building apk file...
> I: Copying unknown files/dir...
> I: Built apk...

Problem solved. You might need to use a new key to sign this file to get it to work on your phone, but it decompiles and recompiles without any issue.

But there's a second part to this. As @JimDunn1234 and @iBotPeaches discussed, something happens to your "**drawable-de-hdpi-v4**" folder. The original APK has a folder called "**drawable-de-hdpi-v4**" but the recompiled APK doesn't. Apktool renames it to "**drawable-de-hdpi**".

In fact there are a lot of folders in that APK that has this happen to. I've taken a screenshot comparing the original APK to the recompiled one, which I've attached.
![APK Comparison Screenshot](https://user-images.githubusercontent.com/18163767/71556890-f88f5e00-2a35-11ea-9f65-e09d3da4ca4c.png)

This seems bad but it actually isn't. The idea behind is that an APK can have, for example, two folders: "**drawable-de-hdpi**" and "**drawable-de-hdpi-v4**". The images in the first folder would normally be used (on a German device with an high screen density) but the images in the second folder would be used if the device is running Android API 4 or above. That's Android 1.4 which I'm reasonably confident you're not using. 
I may be wrong about this, and I'm happy for somebody to correct me, but the fact that this folder has the "**v4**" removed won't realistically make any difference. There's no "**drawable-de-hdpi**" folder in the original APK so nothing's being lost.

So in conclusion, before I start writing an entire damn essay here (I tend to lecture if I'm not paying attention), there's nothing to worry about. The file decompiles and recompiles without any problems.

**Comment by aktivx** [comment]
well i get an error on recompiling the app also, kind of
:brut.androlib.AndrolibException: brut.common.BrutException: could not exec (exit code = -1073741515):
now i have to investigate why you can build it.
it might be that i used a lineage 14.1 framework, or windows 7 x86 with java8 32bit, it did give an error using both aapt and aapt2

**Comment by Ticklefish** [comment]
That exit code is caused by the use of 32 bit Java and shouldn't be happening with apktool 2.4.1. Apktool 2.4.0. had an issue with 32 bit Java which was fixed for the latest version.

**Comment by iBotPeaches** [comment]
Not sure even what the root issue with this report was. 32 bit is deprecated, dying and not supported in the future - so ignoring any of those.

The output in original issue is just warnings, which may point at an issue with our 9patch disassembler, but nothing references an issue with those. 

The `$` sign issue was detected and fixed in the upcoming release.

The remap/char/windows issue is detailed here - https://github.com/iBotPeaches/Apktool/issues/2237#issuecomment-562550340

So closing after breaking down each point in this ticket.
